File: benchmarks/primes.py
import time

limit = 10000000
sieve = bytearray(limit + 1)
t0 = time.monotonic_ns()
for i in range(2, int(limit ** 0.5) + 1):
    if not sieve[i]:
        step = i
        start = i * i
        sieve[start:limit + 1:step] = b'\x01' * ((limit - start) // step + 1)
count = sum(1 for i in range(2, limit + 1) if not sieve[i])
t1 = time.monotonic_ns()
print(f"primes up to 10M: {count}")
print(f"time: {t1 - t0} ns")

File: benchmarks/primes.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <stdint.h>
#include <string.h>

int main() {
    int64_t limit = 10000000;
    char* sieve = (char*)calloc(limit + 1, 1);
    struct timespec ts0, ts1;
    clock_gettime(CLOCK_MONOTONIC, &ts0);
    for (int64_t i = 2; i * i <= limit; i++)
        if (!sieve[i])
            for (int64_t j = i * i; j <= limit; j += i)
                sieve[j] = 1;
    int64_t count = 0;
    for (int64_t i = 2; i <= limit; i++)
        if (!sieve[i]) count++;
    clock_gettime(CLOCK_MONOTONIC, &ts1);
    int64_t elapsed = (int64_t)(ts1.tv_sec - ts0.tv_sec) * 1000000000 +
                      (int64_t)(ts1.tv_nsec - ts0.tv_nsec);
    printf("primes up to 10M: %ld\n", (long)count);
    printf("time: %ld ns\n", (long)elapsed);
    free(sieve);
    return 0;
}
